[GLIB] Skip run loop observer notifications earlier for activities with no observers registered

carlosgcampos opened the pull request:
#### 098750050ead7a58b0424eb9113fa554be8bafb7
<pre>
[GLIB] Skip run loop observer notifications earlier for activities with no observers registered
<a href="https://bugs.webkit.org/show_bug.cgi?id=302392">https://bugs.webkit.org/show_bug.cgi?id=302392</a>

Reviewed by Nikolas Zimmermann.

We currently don&apos;t use Activity::AfterWaiting but on every run loop
iteration we iterate the observers again after waiting for sources to
check if there&apos;s any to notify. We could keep a count of observers
registered for every activity and early return from
RunLoop::notifyActivity when the given activity doesn&apos;t have observers
registered. This patch adds a new class OptionCountedSet with the
minimum API to help with this.

Tests: Tools/TestWebKitAPI/Tests/WTF/OptionCountedSet.cpp
Canonical link: <a href="https://commits.webkit.org/303110@main">https://commits.webkit.org/303110@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2acb23c0d29aad895704e0d13dd4ae067e36f747

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130630 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2901 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41585 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138051 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82253 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/132501 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2913 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2793 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99523 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/67384 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133577 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2113 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116970 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80230 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2034 "Passed tests") | [⏳ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-API-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81307 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/122633 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110610 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35104 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140529 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/129083 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2691 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35606 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108028 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2735 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2439 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107961 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2077 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113314 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55674 "Hash 2acb23c0 for PR 53803 does not build (failure)") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20418 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2761 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31741 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/162098 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2580 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66150 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40426 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2782 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2687 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WTF/wtf/OptionCountedSet.h:
```diff
@@ -0,0 +1,91 @@
+
+namespace WTF {
+
+template<typename E> class OptionCountedSet {
```
How about:

```suggestion
template<typename E> class CountedOptionSet {
```

carlosgcampos replied:
I followed the pattern of HashCountedSet

mcatanzaro commented on Source/WTF/wtf/OptionCountedSet.h:
```diff
@@ -0,0 +1,91 @@
+namespace WTF {
+
+template<typename E> class OptionCountedSet {
+    static_assert(std::is_enum<E>::value, "T is not an enum type");
```
I think this is a typo in OptionSet.h. Surely it should be "E is not an enum type," right?

This static assert is also not necessary since you can't create an `OptionCountedSet<E>` without also reifying `OptionSet<E>`, which will perform the same static assert. But it's fine to be explicit if you prefer, so OK.

carlosgcampos replied:
Oops, I forgot to change this, I'll do it in a follow up, sorry.

nikolaszimmermann reviewed (approved):
LGTM.

carlosgcampos committed:
[GLIB] Skip run loop observer notifications earlier for activities with no observers registered
https://bugs.webkit.org/show_bug.cgi?id=302392

Reviewed by Nikolas Zimmermann.

We currently don't use Activity::AfterWaiting but on every run loop
iteration we iterate the observers again after waiting for sources to
check if there's any to notify. We could keep a count of observers
registered for every activity and early return from
RunLoop::notifyActivity when the given activity doesn't have observers
registered. This patch adds a new class OptionCountedSet with the
minimum API to help with this.

Tests: Tools/TestWebKitAPI/Tests/WTF/OptionCountedSet.cpp
Canonical link: https://commits.webkit.org/303110@main
